src/ipc/client.rs

```diff
@@ -1,102 +1,98 @@
+            // niri instance. Request the running instance version to compare and print a message.
+            Socket::connect()
+                .and_then(|socket| socket.send(Request::Version))
+                .ok()
+        }
+        _ => None,
+    };
 
-    let response = match reply {
-        Ok(r) => r,
-        Err(err_msg) => {
-            eprintln!("The compositor returned an error:");
-            eprintln!();
-            eprintln!("{err_msg}");
+    // Default SIGPIPE so that our prints don't panic on stdout closing.
+    unsafe {
+        libc::signal(libc::SIGPIPE, libc::SIG_DFL);
+    }
 
-            if matches!(msg, Msg::Version) {
-                eprintln!();
-                eprintln!("Note: unable to get the compositor's version.");
-                eprintln!("Did you forget to restart niri after an update?");
-            } else {
-                match NiriSocket::new().and_then(|client| client.send(Request::Version)) {
-                    Ok(Ok(Response::Version(server_version))) => {
-                        let my_version = version();
-                        if my_version != server_version {
-                            eprintln!();
-                            eprintln!("Note: niri msg was invoked with a different version of niri than the running compositor.");
-                            eprintln!("niri msg: {my_version}");
-                            eprintln!("compositor: {server_version}");
-                            eprintln!("Did you forget to restart niri after an update?");
-                        }
-                    }
-                    Ok(Ok(_)) => {
-                        // nonsensical response, do not add confusing context
-                    }
-                    Ok(Err(_)) => {
-                        eprintln!();
-                        eprintln!("Note: unable to get the compositor's version.");
-                        eprintln!("Did you forget to restart niri after an update?");
-                    }
-                    Err(_) => {
-                        // communication error, do not add irrelevant context
-                    }
+    let response = reply.map_err(|err_msg| {
+        // Check for CLI-server version mismatch to add helpful context.
+        match compositor_version {
+            Some(Ok(Response::Version(compositor_version))) => {
+                let cli_version = version();
+                if cli_version != compositor_version {
+                    eprintln!("Running niri compositor has a different version from the niri CLI:");
+                    eprintln!("Compositor version: {compositor_version}");
+                    eprintln!("CLI version:        {cli_version}");
+                    eprintln!("Did you forget to restart niri after an update?");
+                    eprintln!();
                 }
             }
-
-            return Ok(());
+            Some(_) => {
```

Comment:
This block now handles the `Reply::Ok` case for non-`Response::Version`. Is that desirable? This wouldn't be caused by an outdated compositor (it would cause a `Reply::Err("error parsing request")`), and as such i made it not print any kind of error response. In reality, it should be unreachable since the compositor server Does Not Behave Like This

Reply:
Yeah, it is intentional for the reason that you mentioned. If it ever happens then something weird is going on; doesn't hurt to show the warning.